Jenkinsfile

```diff
@@ -901,78 +902,25 @@ pipeline {
-                            //   script '''set -ex
-                            sh script: 'ci/unit/test_post_always.sh',
-                               label: "Collect artifacts and tear down"
-                            junit 'test_results/*.xml'
-                            archiveArtifacts artifacts: 'run_test.sh/**'
-                            archiveArtifacts artifacts: 'vm_test/**'
-                            publishValgrind (
-                                    failBuildOnInvalidReports: true,
-                                    failBuildOnMissingReports: true,
-                                    failThresholdDefinitelyLost: '0',
-                                    failThresholdInvalidReadWrite: '0',
-                                    failThresholdTotal: '0',
-                                    pattern: 'dnt.*.memcheck.xml',
-                                    publishResultsForAbortedBuilds: false,
-                                    publishResultsForFailedBuilds: true,
-                                    sourceSubstitutionPaths: '',
-                                    unstableThresholdDefinitelyLost: '0',
-                                    unstableThresholdInvalidReadWrite: '0',
-                                    unstableThresholdTotal: '0'
-                            )
-                            recordIssues enabledForFailure: true,
-                                         failOnError: true,
-                                         referenceJobName: 'daos-stack/daos/master',
-                                         ignoreFailedBuilds: false,
-                                         ignoreQualityGate: true,
-                                         /* Set qualitygate to 1 new "NORMAL" priority message
-                                           * Supporting messages to help identify causes of
-                                           * problems are set to "LOW", and there are a
-                                           * number of intermittent issues during server
-                                           * shutdown that would normally be NORMAL but in
-                                           * order to have stable results are set to LOW.
-                                           */
-                                         qualityGates: [[threshold: 1, type: 'TOTAL_HIGH', unstable: true],
-                                                        [threshold: 1, type: 'TOTAL_ERROR', unstable: true],
-                                                        [threshold: 1, type: 'NEW_NORMAL', unstable: true]],
-                                         name: "Node local testing",
-                                         tool: issues(pattern: 'vm_test/nlt-errors.json',
-                                                      name: 'NLT results',
-                                                      id: 'VM_test')
+                            unitTestPost()
```

Comment:
It's great that all this code is being put in functions so it's not in-line with the rest of the pipeline here, but having it in a separate repo seems like a retrograde step to me, I can't figure out what the benefit of that is at all.

Reply:
@ashleypittman It does seem odd for the `Jenkinsfile` in the daos repo to call a function in the pipeline-lib repo which just then ends up executing a script back in the daos repo again.  I also considered whether we needed the extra level of indirection (through pipline-lib) but we do.  This is how we condense down many statements in the `Jenkinsfile` into a single statement, allowing us to have more stages in `Jenkinsfile` before hitting the limit in the size of `Jenkinsfile`.

Reply:
Unfortunately the way that Jenkins currently loads the pipeline libraries pretty much requires them to be in a separate repository than the routine that uses them.

Jenkins does not have a way to always load a library from checkout of the repository being built.   This causes some issues with the Unit testing that we are adding to the pipeline libraries.

Reply:
I don't understand how this could be an issue.   The Jenkinsfile is used from the repo being tested so wouldn't that mean things are checked out?   There is no way to load a script from a relative path ?

Reply:
The SCM checkout of the library scripts are different checkouts than the code run for the PR.

When a pipeline Jenkins jog is triggered Jenkins does the following:

1. Checks out the Jenkinsfile after determining if the trigger is allowed to use the Jenkinsfile from the current branch or is restricted to master.

2. Checks out the groovy libraries requested by the system, github project settings, and Jenkinsfile used.

3. Checks out the SCM code to be built.   This is only code visible in the workspace.

The specific branch of the groovy script library to be used is set in one of two places.

1. In the github project setup inside Jenkins.  This library is common to all repositories in the github project.

2. With the `@Library(value="library-name@library_branch") _` specified in the Jenkinsfile being used.

If you set the Jenkinsfile to use "daos@my_pr_branch" in a PR, then that is what it will use.  If you merge in that patch
with out changing the Jenkinsfile to "daos@master" then master will use that branch and if that branch no longer exists, master will fail.

There is no way that we have found to tell Jenkins to use a library that is always at the same branch as being built.